src/gui/log/logmodel.cpp

```diff
@@ -0,0 +1,192 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "logmodel.h"
+
+#include <QApplication>
+#include <QDateTime>
+#include <QColor>
+#include <QPalette>
+
+#include "base/global.h"
+
+namespace
+{
+    const int MAX_VISIBLE_MESSAGES = 20000;
+}
+
+BaseLogModel::Message::Message(const QString &time, const QString &message, const QColor &foreground, const Log::MsgType type)
+    : m_time(time)
+    , m_message(message)
+    , m_foregroundRole(foreground)
+    , m_userRole(type)
+{
+}
+
+QVariant BaseLogModel::Message::time() const
+{
+    return m_time;
+}
+
+QVariant BaseLogModel::Message::message() const
+{
+    return m_message;
+}
+
+QVariant BaseLogModel::Message::foregroundRole() const
+{
+    return m_foregroundRole;
+}
+
+QVariant BaseLogModel::Message::userRole() const
+{
+    return m_userRole;
+}
+
+BaseLogModel::BaseLogModel(QObject *parent)
+    : QAbstractListModel(parent)
+    , m_messages(MAX_VISIBLE_MESSAGES)
+{
+}
+
+int BaseLogModel::rowCount(const QModelIndex &) const
+{
+    return m_messages.size();
+}
+
+int BaseLogModel::columnCount(const QModelIndex &) const
+{
+    return 1;
+}
+
+QVariant BaseLogModel::data(const QModelIndex &index, const int role) const
+{
+    if (!index.isValid())
+        return {};
+
+    const int messageIndex = index.row();
+    if ((messageIndex < 0) || (messageIndex >= m_messages.size()))
+        return {};
+
+    const Message &message = m_messages[messageIndex];
+    switch (role) {
+    case Qt::ForegroundRole:
+        return message.foregroundRole();
+    case Qt::UserRole:
+        return message.userRole();
+    case TimeRole:
+        return message.time();
+    case MessageRole:
+        return message.message();
+    default:
+        return {};
+    }
+}
+
+void BaseLogModel::addNewMessage(const BaseLogModel::Message &message)
+{
+    // if row is inserted on filled up buffer, the size will not change but because of calling of beginInsertRows function
```

Comment:
Please reformat this comment. The first line is too long while the secnd one is short.